File: examples/go-feature-sliced/order/domain/order.go
package domain

import (
	"fmt"

	shareddomain "github.com/example/ecommerce/shared/domain"
)

// OrderStatus represents the lifecycle state of an order.
// Corresponds to: model OrderStatus : enterprise (enum)
type OrderStatus string

const (
	StatusPending   OrderStatus = "Pending"
	StatusConfirmed OrderStatus = "Confirmed"
	StatusShipped   OrderStatus = "Shipped"
	StatusDelivered OrderStatus = "Delivered"
	StatusCancelled OrderStatus = "Cancelled"
)

// OrderItem is a value object within an Order.
// Corresponds to: model OrderItem : enterprise (value object)
type OrderItem struct {
	ProductID string
	Name      string
	Quantity  int
	Price     shareddomain.Money
}

func (item OrderItem) Subtotal() shareddomain.Money {
	return shareddomain.Money{
		Amount:   item.Price.Amount * float64(item.Quantity),
		Currency: item.Price.Currency,
	}
}

// Order is the aggregate root.
// Corresponds to: model Order : enterprise (aggregate)
//
//	invariant positiveTotal: total.amount > 0
type Order struct {
	ID         string
	CustomerID string
	Items      []OrderItem
	Total      shareddomain.Money
	Shipping   shareddomain.Address
	Status     OrderStatus
}

// Validate checks the positiveTotal invariant.
func (o *Order) Validate() error {
	if o.Total.Amount <= 0 {
		return fmt.Errorf("invariant violated: total must be positive, got %f", o.Total.Amount)
	}
	return nil
}

func (o *Order) Cancel() error {
	if o.Status == StatusCancelled {
		return fmt.Errorf("order already cancelled")
	}
	if o.Status == StatusDelivered {
		return fmt.Errorf("cannot cancel delivered order")
	}
	o.Status = StatusCancelled
	return nil
}

func (o *Order) Confirm() {
	o.Status = StatusConfirmed
}

// RecalculateTotal computes total from items.
func (o *Order) RecalculateTotal() {
	var total float64
	currency := "USD"
	for _, item := range o.Items {
		sub := item.Subtotal()
		total += sub.Amount
		currency = sub.Currency
	}
	o.Total = shareddomain.Money{Amount: total, Currency: currency}
}

File: examples/go-feature-sliced/architecture/Order.hs
module Order (
  orderStatus, order, orderItem,
  orderRepo,
  placeOrder, cancelOrder, getOrder, listOrders,
  memOrderRepo,
  orderModule,
  declareAll
) where

import Plat.Core
import Plat.Ext.CleanArch
import Plat.Ext.DDD
import Plat.Ext.Modules
import Shared (money, address)
import Payment (paymentGateway, paymentModule)

----------------------------------------------------------------------
-- Feature: Order
----------------------------------------------------------------------

-- Domain

orderStatus :: Decl 'Model
orderStatus = enum "OrderStatus" enterprise
  ["Pending", "Confirmed", "Shipped", "Delivered", "Cancelled"]

order :: Decl 'Model
order = aggregate "Order" enterprise $ do
  field "id"        (customType "UUID")
  field "customerId" (customType "UUID")
  field "items"     (listOf orderItem)
  field "total"     (ref money)
  field "shipping"  (ref address)
  field "status"    (ref orderStatus)
  invariant "positiveTotal" "total.amount > 0"

orderItem :: Decl 'Model
orderItem = value "OrderItem" enterprise $ do
  field "productId" (customType "UUID")
  field "name"      string
  field "quantity"  int
  field "price"     (ref money)

-- Port

orderRepo :: Decl 'Boundary
orderRepo = port "OrderRepository" interface $ do
  op "save"     ["order" .: ref order] ["err" .: error_]
  op "findById" ["id" .: customType "UUID"] ["order" .: ref order, "err" .: error_]
  op "findAll"  [] ["orders" .: listOf order, "err" .: error_]
  op "delete"   ["id" .: customType "UUID"] ["err" .: error_]

-- Use cases

placeOrder :: Decl 'Operation
placeOrder = usecase "PlaceOrder" application $ do
  input  "order"       (ref order)
  input  "paymentToken" string
  output "orderId"     (customType "UUID")
  output "err"         error_
  needs orderRepo
  needs paymentGateway

cancelOrder :: Decl 'Operation
cancelOrder = usecase "CancelOrder" application $ do
  input  "orderId" (customType "UUID")
  output "err"     error_
  needs orderRepo

getOrder :: Decl 'Operation
getOrder = usecase "GetOrder" application $ do
  input  "orderId" (customType "UUID")
  output "order"   (ref order)
  output "err"     error_
  needs orderRepo

listOrders :: Decl 'Operation
listOrders = usecase "ListOrders" application $ do
  output "orders" (listOf order)
  output "err"    error_
  needs orderRepo

-- Adapter

memOrderRepo :: Decl 'Adapter
memOrderRepo = impl "InMemoryOrderRepo" framework orderRepo $ do
  inject "store" (ext "sync.Map")

-- Module

orderModule :: Decl 'Compose
orderModule = domain "OrderFeature" $ do
  import_ paymentModule paymentGateway
  expose order
  expose orderItem
  expose orderStatus
  expose orderRepo
  expose placeOrder
  expose cancelOrder
  expose getOrder
  expose listOrders
  expose memOrderRepo

declareAll :: ArchBuilder ()
declareAll = declares
  [ decl orderStatus
  , decl order
  , decl orderItem
  , decl orderRepo
  , decl placeOrder
  , decl cancelOrder
  , decl getOrder
  , decl listOrders
  , decl memOrderRepo
  , decl orderModule
  ]
